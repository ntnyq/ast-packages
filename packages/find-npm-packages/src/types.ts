import type { ParserOptions } from '@babel/parser'

export interface NpmPackage {
  end: number | null
  name: string
  start: number | null
  loc: {
    end: Position
    start: Position
  }
}

export interface Options extends ParserOptions {
  /**
   * Parse code cache
   */
  cache?: boolean

  /**
   * Code language
   *
   * @default `js`
   */
  language?: 'dts' | 'js' | 'jsx' | 'ts'
}

interface Position {
  column: number
  index: number
  line: number
}
