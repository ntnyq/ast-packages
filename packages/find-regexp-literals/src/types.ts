import type { ParserOptions } from '@babel/parser'

export interface Options extends ParserOptions {
  /**
   * Parse code cache
   */
  cache?: boolean

  /**
   * Code language
   *
   * @default `js`
   */
  language?: 'dts' | 'js' | 'jsx' | 'ts'
}

export interface RegExpLiteral {
  flags: string
  pattern: string
  end?: number | null
  start?: number | null
  loc?: {
    end?: Position
    start?: Position
  }
}

interface Position {
  column: number
  index: number
  line: number
}
