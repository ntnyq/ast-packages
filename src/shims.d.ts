declare module 'validate-npm-package-name' {
  interface Results {
    validForNewPackages: boolean
    validForOldPackages: boolean
    errors?: string[] | undefined
    warnings?: string[] | undefined
  }

  interface ValidNames extends Results {
    validForNewPackages: true
    validForOldPackages: true
  }

  interface InvalidNames extends Results {
    errors: string[]
    validForNewPackages: false
    validForOldPackages: false
  }

  interface LegacyNames extends Results {
    validForNewPackages: false
    validForOldPackages: true
    warnings: string[]
  }

  declare function validate(
    name: string,
  ): InvalidNames | LegacyNames | ValidNames

  export = validate
}
