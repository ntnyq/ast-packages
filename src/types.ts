import type { ParserOptions } from '@babel/parser'

export interface NpmPackage {
  name: string
  end?: number | null
  start?: number | null
  loc?: {
    end?: Position
    start?: Position
  }
}

export interface Options extends ParserOptions {
  /**
   * Parse code cache
   */
  cache?: boolean

  /**
   * Code language
   *
   * @default `js`
   */
  language?: 'dts' | 'js' | 'jsx' | 'ts'
}

interface Position {
  column: number
  index: number
  line: number
}
