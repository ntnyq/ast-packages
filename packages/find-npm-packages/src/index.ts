import { babelParse, walkAST } from 'ast-kit'
import validateNpmPackageName from 'validate-npm-package-name'
import type * as t from '@babel/types'
import type { NpmPackage, Options } from './types'

const RE_PACKAGE_NAME = /^(@[^/]+\/[^/@]+|[^/@]+)/

/**
 * Find npm packages from given source code.
 *
 * @param code - Code to parse
 * @param options - options
 * @returns An array of found npm packages
 *
 * @example
 *
 * ```ts
 * import { findNpmPackages } from 'find-npm-packages'
 *
 * const code = `import findNpmPackages from 'find-npm-packages'`
 *
 * const result = findNpmPackages(code)
 * ```
 */
export function findNpmPackages(
  code: string,
  options: Options = {},
): NpmPackage[] {
  const { language = 'js' } = options
  const program = babelParse(code, language, options)
  const packages: NpmPackage[] = []

  function getValidNpmPackageName(importString: string) {
    const match = importString.match(RE_PACKAGE_NAME)

    if (match) {
      const name = match[0]
      const result = validateNpmPackageName(name)

      if (result.validForNewPackages || result.validForOldPackages) {
        return name
      }
    }
  }

  function checkNpmPackageName(stringLiteral: t.StringLiteral) {
    const name = getValidNpmPackageName(stringLiteral.value)

    if (!name) return

    const start = stringLiteral.start
    const loc = stringLiteral.loc

    if (start && loc) {
      packages.push({
        end: start + name.length + 1,
        name,
        start: start + 1,
        loc: {
          end: {
            column: loc.start.column + name.length + 1,
            index: loc.start.index + name.length + 1,
            line: loc.start.line,
          },
          start: {
            column: loc.start.column + 1,
            index: loc.start.index + 1,
            line: loc.start.line,
          },
        },
      })
    }
  }

  walkAST(program, {
    enter(node) {
      if (node.type === 'ImportDeclaration') {
        checkNpmPackageName(node.source)
      }

      // NOTE: ImportExpression is not supported
      if (node.type === 'CallExpression') {
        const isImportExpression = node.callee.type === 'Import'
        const isRequireExpression =
          node.callee.type === 'Identifier' && node.callee.name === 'require'

        if (
          (isImportExpression || isRequireExpression)
          && node.arguments[0]
          && node.arguments[0].type === 'StringLiteral'
        ) {
          checkNpmPackageName(node.arguments[0])
        }
      }
    },
  })

  return packages
}
