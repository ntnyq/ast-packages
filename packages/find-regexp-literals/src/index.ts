import { babelParse, walkAST } from 'ast-kit'
import type { Options, RegExpLiteral } from './types'

export function findRegExpLiterals(
  code: string,
  options: Options = {},
): RegExpLiteral[] {
  const { language = 'js' } = options
  const program = babelParse(code, language, options)
  const result: RegExpLiteral[] = []

  walkAST(program, {
    enter(node) {
      if (node.type === 'RegExpLiteral' || node.type === 'RegexLiteral') {
        result.push({
          end: node.end,
          flags: node.flags,
          pattern: node.pattern,
          start: node.start,
          loc: {
            ...node.loc,
          },
        })
      }
    },
  })

  return result
}
