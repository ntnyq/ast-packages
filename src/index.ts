import { babelParse, walkAST } from 'ast-kit'
import validateNpmPackageName from 'validate-npm-package-name'
import type { NpmPackage, Options } from './types'

/**
 * Find npm packages from given source code.
 *
 * @param code - Code to parse
 * @param options - options
 * @returns An array of found npm packages
 *
 * @example
 *
 * ```ts
 * import { findNpmPackages } from 'find-npm-packages'
 *
 * const code = `import findNpmPackages from 'find-npm-packages'`
 *
 * const result = findNpmPackages(code)
 * ```
 */
export function findNpmPackages(
  code: string,
  options: Options = {},
): NpmPackage[] {
  const { language = 'js' } = options
  const program = babelParse(code, language, options)
  const packages: NpmPackage[] = []

  function isNpmPackageName(name: string) {
    const result = validateNpmPackageName(name)
    return result.validForNewPackages || result.validForOldPackages
  }

  walkAST(program, {
    enter(node) {
      if (
        node.type === 'ImportDeclaration'
        && isNpmPackageName(node.source.value)
      ) {
        packages.push({
          end: node.source.end,
          name: node.source.value,
          start: node.source.start,
          loc: {
            end: node.source.loc?.end,
            start: node.source.loc?.start,
          },
        })
      }

      // NOTE: ImportExpression is not supported
      if (node.type === 'CallExpression') {
        const isImportExpression = node.callee.type === 'Import'
        const isRequireExpression =
          node.callee.type === 'Identifier' && node.callee.name === 'require'

        if (
          (isImportExpression || isRequireExpression)
          && node.arguments[0]
          && node.arguments[0].type === 'StringLiteral'
          && isNpmPackageName(node.arguments[0].value)
        ) {
          packages.push({
            end: node.arguments[0].end,
            name: node.arguments[0].value,
            start: node.arguments[0].start,
            loc: {
              end: node.arguments[0].loc?.end,
              start: node.arguments[0].loc?.start,
            },
          })
        }
      }
    },
  })

  return packages
}
