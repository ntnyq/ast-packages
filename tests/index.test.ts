import { expect, it } from 'vitest'
import { findNpmPackages } from '../src'
import FIXTURE_DTS from './fixtures/dts.d.ts?raw'
import FIXTURE_JS from './fixtures/js.js?raw'
import FIXTURE_JSX from './fixtures/jsx.jsx?raw'
import FIXTURE_TS from './fixtures/ts.ts?raw'

it('basic', () => {
  expect(findNpmPackages(`import findNpmPackages from 'find-npm-packages'`))
    .toMatchInlineSnapshot(`
    [
      {
        "end": 47,
        "loc": {
          "end": Position {
            "column": 47,
            "index": 47,
            "line": 1,
          },
          "start": Position {
            "column": 28,
            "index": 28,
            "line": 1,
          },
        },
        "name": "find-npm-packages",
        "start": 28,
      },
    ]
  `)
})

it('js', () => {
  expect(
    findNpmPackages(FIXTURE_JS.replace(/\r\n/g, '\n'), {
      language: 'js',
    }),
  ).toMatchSnapshot()
})

it('ts', () => {
  expect(
    findNpmPackages(FIXTURE_TS.replace(/\r\n/g, '\n'), {
      language: 'ts',
    }),
  ).toMatchSnapshot()
})

it('dts', () => {
  expect(
    findNpmPackages(FIXTURE_DTS.replace(/\r\n/g, '\n'), {
      language: 'dts',
    }),
  ).toMatchSnapshot()
})

it('jsx', () => {
  expect(
    findNpmPackages(FIXTURE_JSX.replace(/\r\n/g, '\n'), {
      language: 'jsx',
    }),
  ).toMatchSnapshot()
})
