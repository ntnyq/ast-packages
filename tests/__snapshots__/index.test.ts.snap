// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`dts 1`] = `
[
  {
    "end": 37,
    "loc": {
      "end": Position {
        "column": 21,
        "index": 37,
        "line": 3,
      },
      "start": Position {
        "column": 16,
        "index": 32,
        "line": 3,
      },
    },
    "name": "foo",
    "start": 32,
  },
  {
    "end": 65,
    "loc": {
      "end": Position {
        "column": 27,
        "index": 65,
        "line": 4,
      },
      "start": Position {
        "column": 17,
        "index": 55,
        "line": 4,
      },
    },
    "name": "@foo/foo",
    "start": 55,
  },
  {
    "end": 97,
    "loc": {
      "end": Position {
        "column": 31,
        "index": 97,
        "line": 5,
      },
      "start": Position {
        "column": 22,
        "index": 88,
        "line": 5,
      },
    },
    "name": "foo-foo",
    "start": 88,
  },
  {
    "end": 128,
    "loc": {
      "end": Position {
        "column": 30,
        "index": 128,
        "line": 6,
      },
      "start": Position {
        "column": 21,
        "index": 119,
        "line": 6,
      },
    },
    "name": "foo.foo",
    "start": 119,
  },
  {
    "end": 167,
    "loc": {
      "end": Position {
        "column": 38,
        "index": 167,
        "line": 7,
      },
      "start": Position {
        "column": 25,
        "index": 154,
        "line": 7,
      },
    },
    "name": "foo-foo-foo",
    "start": 154,
  },
]
`;

exports[`js 1`] = `
[
  {
    "end": 21,
    "loc": {
      "end": Position {
        "column": 21,
        "index": 21,
        "line": 1,
      },
      "start": Position {
        "column": 16,
        "index": 16,
        "line": 1,
      },
    },
    "name": "foo",
    "start": 16,
  },
  {
    "end": 49,
    "loc": {
      "end": Position {
        "column": 27,
        "index": 49,
        "line": 2,
      },
      "start": Position {
        "column": 17,
        "index": 39,
        "line": 2,
      },
    },
    "name": "@foo/foo",
    "start": 39,
  },
  {
    "end": 81,
    "loc": {
      "end": Position {
        "column": 31,
        "index": 81,
        "line": 3,
      },
      "start": Position {
        "column": 22,
        "index": 72,
        "line": 3,
      },
    },
    "name": "foo-foo",
    "start": 72,
  },
  {
    "end": 112,
    "loc": {
      "end": Position {
        "column": 30,
        "index": 112,
        "line": 4,
      },
      "start": Position {
        "column": 21,
        "index": 103,
        "line": 4,
      },
    },
    "name": "foo.foo",
    "start": 103,
  },
  {
    "end": 216,
    "loc": {
      "end": Position {
        "column": 25,
        "index": 216,
        "line": 9,
      },
      "start": Position {
        "column": 20,
        "index": 211,
        "line": 9,
      },
    },
    "name": "bar",
    "start": 211,
  },
  {
    "end": 274,
    "loc": {
      "end": Position {
        "column": 24,
        "index": 274,
        "line": 12,
      },
      "start": Position {
        "column": 19,
        "index": 269,
        "line": 12,
      },
    },
    "name": "baz",
    "start": 269,
  },
]
`;

exports[`jsx 1`] = `
[
  {
    "end": 21,
    "loc": {
      "end": Position {
        "column": 21,
        "index": 21,
        "line": 1,
      },
      "start": Position {
        "column": 16,
        "index": 16,
        "line": 1,
      },
    },
    "name": "foo",
    "start": 16,
  },
  {
    "end": 49,
    "loc": {
      "end": Position {
        "column": 27,
        "index": 49,
        "line": 2,
      },
      "start": Position {
        "column": 17,
        "index": 39,
        "line": 2,
      },
    },
    "name": "@foo/foo",
    "start": 39,
  },
  {
    "end": 81,
    "loc": {
      "end": Position {
        "column": 31,
        "index": 81,
        "line": 3,
      },
      "start": Position {
        "column": 22,
        "index": 72,
        "line": 3,
      },
    },
    "name": "foo-foo",
    "start": 72,
  },
  {
    "end": 112,
    "loc": {
      "end": Position {
        "column": 30,
        "index": 112,
        "line": 4,
      },
      "start": Position {
        "column": 21,
        "index": 103,
        "line": 4,
      },
    },
    "name": "foo.foo",
    "start": 103,
  },
  {
    "end": 165,
    "loc": {
      "end": Position {
        "column": 25,
        "index": 165,
        "line": 7,
      },
      "start": Position {
        "column": 20,
        "index": 160,
        "line": 7,
      },
    },
    "name": "bar",
    "start": 160,
  },
  {
    "end": 223,
    "loc": {
      "end": Position {
        "column": 24,
        "index": 223,
        "line": 10,
      },
      "start": Position {
        "column": 19,
        "index": 218,
        "line": 10,
      },
    },
    "name": "baz",
    "start": 218,
  },
]
`;

exports[`ts 1`] = `
[
  {
    "end": 37,
    "loc": {
      "end": Position {
        "column": 21,
        "index": 37,
        "line": 3,
      },
      "start": Position {
        "column": 16,
        "index": 32,
        "line": 3,
      },
    },
    "name": "foo",
    "start": 32,
  },
  {
    "end": 65,
    "loc": {
      "end": Position {
        "column": 27,
        "index": 65,
        "line": 4,
      },
      "start": Position {
        "column": 17,
        "index": 55,
        "line": 4,
      },
    },
    "name": "@foo/foo",
    "start": 55,
  },
  {
    "end": 97,
    "loc": {
      "end": Position {
        "column": 31,
        "index": 97,
        "line": 5,
      },
      "start": Position {
        "column": 22,
        "index": 88,
        "line": 5,
      },
    },
    "name": "foo-foo",
    "start": 88,
  },
  {
    "end": 128,
    "loc": {
      "end": Position {
        "column": 30,
        "index": 128,
        "line": 6,
      },
      "start": Position {
        "column": 21,
        "index": 119,
        "line": 6,
      },
    },
    "name": "foo.foo",
    "start": 119,
  },
  {
    "end": 167,
    "loc": {
      "end": Position {
        "column": 38,
        "index": 167,
        "line": 7,
      },
      "start": Position {
        "column": 25,
        "index": 154,
        "line": 7,
      },
    },
    "name": "foo-foo-foo",
    "start": 154,
  },
  {
    "end": 220,
    "loc": {
      "end": Position {
        "column": 25,
        "index": 220,
        "line": 10,
      },
      "start": Position {
        "column": 20,
        "index": 215,
        "line": 10,
      },
    },
    "name": "bar",
    "start": 215,
  },
  {
    "end": 278,
    "loc": {
      "end": Position {
        "column": 24,
        "index": 278,
        "line": 13,
      },
      "start": Position {
        "column": 19,
        "index": 273,
        "line": 13,
      },
    },
    "name": "baz",
    "start": 273,
  },
]
`;
